### ISSUE TITLE
CMS integration tracking

### ISSUE THREAD

**Issue by Maikuolan** [phpMussel/phpMussel#57]
Following up from https://github.com/Maikuolan/phpMussel/issues/55, where @bechtold has offered to contribute a Drupal module based on phpMussel, I think the idea of integrating phpMussel with various CMS is a great idea, if it could be done.

I haven't yet seen any offers from anybody else to integrate phpMussel elsewhere, though, if anybody reads this issue and thinks that they might be interested in helping out, I'd love to hear from them.

I'm creating this issue to track and cross-reference research and communications relating to possible future integration of phpMussel into various CMS.

Regarding phpBB, via a reply to a support request for phpMussel posted to Spambot Security (linked **[here](http://www.spambotsecurity.com/forum/viewtopic.php?f=59&t=3535&start=40)**), though as of yet, this hasn't been followed up:

> I've been reading the documentation for creating plugins for phpBB (referred to as "**[extensions](https://www.phpbb.com/extensions/)**" for phpBB versions 3.1.x and referred to as "**[MODs](https://www.phpbb.com/mods/)**" for phpBB versions 3.0.x; doesn't appear to be any documentation for phpBB versions lower than 3.x). It should be possible to create a plugin for phpBB, although, extensions and MODs aren't built and don't function in the same way as each other, and so, the two wouldn't be mutually compatible; Depending on which version of phpBB is used, one or the other may be required, or possibly both, depending on circumstances and intended usage. I'm also wondering about the likelihood of the phpBB team bringing out some other new system for plugins when they eventually reach version 3.2.x, and whether any such system would be mutually compatible with extensions (probably not); That might make things much more tedious in the future for maintaining any such plugins.
> 
> Alternatively, _(snip)_ I might possibly be able to just add some new code and associated directive to phpMussel dealing specifically with phpBB compatibility for instructing phpMussel to return its Upload Denied message as a JSON response that could be interpreted by phpBB rather than as an actual HTML page. Something to look into, anyhow.

Regarding phpBB, interest for anti-virus integration for uploads seems to be mixed; Reading their support forums, I've seen this subject discussed numerous times in the past, although there doesn't seem to have been much progress in regards to the subject. I've seen a number of users suggest that they think the idea is silly and unneeded, but at the same time, over the years, it seems many users have actually made requests for either this specific functionality or solutions similar to this functionality.

Searching their forum for terms such as "virus", "malware", "shell" and etc, can reveal this. For example - https://www.phpbb.com/search/?q=virus

A few months ago, I attempted to reply to one such support request, as a means to test the waters for something like this: https://www.phpbb.com/community/viewtopic.php?f=64&t=2312951

There were only two responses - one for, one against - this indicates what I mean by mixed interest. Nonetheless, I think this is definitely something that should be done.

Regarding Wordpress, posted via their support forum (linked **[here](https://wordpress.org/support/topic/upload-plugin-6)**), though unfortunately, at this point in time, there hasn't been any replies:

> Firstly, it has been suggested to me that WordPress leverages AJAX for its file upload process; Is that true?
> 
> Secondly, if the above is true, what would be the appropriate action/hook to use in order to trigger the primary function of the plugin upon a user attempting to upload a file via WordPress?
> 
> Thirdly, and assuming again the above to be true, what response would be expected from the request and how would WordPress handle that response?

**Comment by Wardormeur** [comment]
Just a word : 
MOD by definition are the extensions for phpbb <3.0, whereas extensions are for phpbb3.1+
The logic is a breaking change from the restructuration of the phpbb system in 3.1, MOD were MODifications of the core system, whereas extensions now uses some core events mapping. Noone's gonna make any MOD anymore, for maintainability sake.
On the security side, i agree that another layer would be interested and "not so hard" to implement, however the public of phpbb is far from being technophile, and security the least concern. Plus, the upload are by default very limited (file as attachment), while possible to bypass them ( folder with ext .jpg with an index.php is still "valid" as a bbcode), but still very basic without "real workflow"/usage as the wysywig editors are still on development (or as an ext), so that their concern for the core may be limited, while someone with an heavily customised phpbb should really care :p
Here for their event list where to plug-in : https://wiki.phpbb.com/Event_List

On WP side, it's again the same story : care about your plugins more than the core itself. Same story again, here the events are called "Hooks" (http://codex.wordpress.org/Plugin_API/Action_Reference)
I confirm it's ajax-based, our would recommend to get a look @ http://codex.wordpress.org/Plugin_API/Action_Reference/add_attachment & its alternative, edit_attachment

IDK for the WP side, but the phpbb guys on freenode#phpbb-coding are helpful ;)

**Comment by Maikuolan** [comment]
> MOD by definition are the extensions for phpbb <3.0, whereas extensions are for phpbb3.1+
> The logic is a breaking change from the restructuration of the phpbb system in 3.1, MOD were MODifications of the core system, whereas extensions now uses some core events mapping. Noone's gonna make any MOD anymore, for maintainability sake.

That does make a lot of sense, and in hindsight, seems obvious. I wasn't entirely clear on that, however, so, I appreciate your input. :P

Considering that, I think, now, that I'd be much more interested in seeing an extension developed than in seeing a mod developed.

I was, once upon a time, many many years ago, involved with a few projects for modding phpBB (building simple things like shoutbox modules, IRC integration, those sorts of things), but, that was back when phpBB2 was prevalent, and since that time, I've taken only the slightest glance at the code-base for phpBB, and so, although I'm a user of phpBB3 in the sense that I'm registered to handful of forums that use phpBB3 (and even administrate at one of them at the moment, although everything I've done or I'm doing there was/is from front-end rather than back-end), from a development perspective, I'm almost entirely clueless about it.

Cheers for the links; Was looking for something like those as a point of reference (having that information should save a lot of time). :-)

> IDK for the WP side, but the phpbb guys on freenode#phpbb-coding are helpful ;)

I'll check that out. :-)

**Comment by DanielRuf** [comment]
list of software where an integration of phpMussel makes sense:
- [ ] Drupal
- [ ] WordPress
- [ ] Magento
- [ ] Joomla
- [ ] Contao
- [ ] TYPO3
- [ ] MyBB
- [ ] phpBB
- [ ] XenForo
- [ ] vBulletin
- [ ] WoltLab
- [ ] Simple Machines Forum
- [ ] Vanilla Forums
- [ ] Invision Power Services

**Comment by Maikuolan** [comment]
:+1: :-)

**Comment by Maikuolan** [comment]
Tried asking on the IRC channel last night, but didn't have much luck. :-/

Still trying various different channels though.

**Comment by DanielRuf** [comment]
> Tried asking on the IRC channel last night, but didn't have much luck. :-/

The IRC channel of which listed product / CMS?

**Comment by Maikuolan** [comment]
> The IRC channel of which listed product / CMS?

The phpBB coding IRC channel @ freenode:

> IDK for the WP side, but the phpbb guys on freenode#phpbb-coding are helpful ;)

I don't have anything against their specific channel, and judging from the number of users that were connected to it, I guess that it would probably be considered one of the more popular or successful channels, but personally, I've just never really liked IRC channels very much in general; For me, using IRC feels similar to watching paint dry, and feels like a waste of time: So much waiting around and doing nothing.

Although there are exceptions, over the course of many years of trying to benefit from IRC, most of the IRC channels that I've found and tried to use tend to always be heavily populated by a large number of users that just do nothing; Remain idle, connected, but don't chat or ever say anything. I just think it's strange, that for a system designed to allow people to chat with one another, that most people that use it don't really seem to actually chat to anybody. I've always much preferred things like forums, email, etc.

IRC can be very useful when everything is coordinated beforehand, though, I've found; In the past, when I was involved with some online political roleplaying games, myself and other players would frequently use IRC to plan wars and other things like that that we wanted to keep away from the eyes of enemy players; Was very useful in those circumstances. I haven't really used it much since those days though.

Anyhow, concerning phpBB-coding: I logged on, described the problem we're having here and described what we're looking for; I stayed connected for about 9 hours, and in that time, one user suggested someone else on the channel that might be able to help and tried to get their attention, but the person that they thought might've been able to help didn't respond, and there weren't any other responses to my messages otherwise.

**Comment by Elsensee** [comment]
Hi,

I guess you're talking about that:

```
<Maikuolan> Looking for someone that knows a bit about the AJAX file uploading capabilities / post file attachment functionality of phpBB for a few pointers. If anyone's around that could help, let me know. Cheers.
<Elsensee> prototech: ^ :D
<Maikuolan> If prototech is able to help, that'd be fantastic. Or anyone, for that matter. I'm not picky. :P
<Maikuolan> Starting work soon so going to have a quick rest, but I'll leave my machine on and come back when I'm up again; Will be on most of the night; If anyone is around, just give me a shout, I'll reply at some point during the night.
```

That is very unfortunate and I'm sorry. I just pointed you to him because AFAIK he's the one who implemented plupload into phpBB so he could help you best.

For people who are familiar with OOP in PHP, writing extensions for phpBB shouldn't be very hard. For people who are familiar with Symfony it should be even easier.

Currently it looks like phpBB doesn't have many events to hook into for the upload, but that's not bad - we/you can add them easily.
If you need some assistance regarding phpBB extensions you can PM me on phpBB.com or poke me on IRC. Sometimes there are other users as well who could help you with extension writing - I don't know why they didn't show up on that day.

You might also look into composer integration for your script.

By the waay: A good starting point might be our new phpBB Development Documentation we started writing at our annual hackathon this year: https://area51.phpbb.com/docs/dev/31x/ :)

**Comment by Wardormeur** [comment]
MB open a ticket for phpbb only :)
fileupload is done through the "fileupload" class ($phpbb_root_path . 'includes/functions_upload.php ) But not sure at all it supports ajax upload, i think it's on submit /posting.php || /includes/functions_posting.php

**Comment by Maikuolan** [comment]
@Elsensee:

> That is very unfortunate and I'm sorry. I just pointed you to him because AFAIK he's the one who implemented plupload into phpBB so he could help you best.

Makes sense; And I appreciate the help. :-)

Your replying here, too, shows that you're trying to help; I didn't link back to this repository when I posted those messages to the IRC channel, so, I guess you must've searched my username or something similar to find this repository after I posted those messages; Past experiences have shown me that not many people usually bother to go that far to respond. Anyhow, I appreciate the help. Nothing to be sorry about; Not your fault, and sometimes it takes more than a single try or a single day to find/get help with things like these, of course.

> If you need some assistance regarding phpBB extensions you can PM me on phpBB.com or poke me on IRC. Sometimes there are other users as well who could help you with extension writing - I don't know why they didn't show up on that day.

That would be great. :-)

I'll read through that documentation later today after I get home from work; It looks more thorough than the documentation I'd read before, and could be very useful.

@Wardormeur:

> But not sure at all it supports ajax upload, i think it's on submit /posting.php || /includes/functions_posting.php

Hm, interesting. I'd been told otherwise by someone else. I might need to take a closer look at those files, I guess. Thanks for the tip. :-)

**Comment by Elsensee** [comment]
Actually @DanielRuf posted this on twitter: https://twitter.com/DanielRufde/status/629599479877009408 and I remembered you.
My goal is just that every person asking for help gets their help. :)

**Comment by Maikuolan** [comment]
Ah, okay. Cool. :-)

BTW, @DanielRuf, great work with the twitter posts. :+1:

**Comment by DanielRuf** [comment]
> Ah, okay. Cool. :-)
> 
> BTW, @DanielRuf, great work with the twitter posts.

No problem, I am just helping and contributing as much as I can and trying to contact developers who I know who could help and contribute to phpMussel. Unfortunately I have not worked with phpBB before and do not know its internals or API structure.

**Comment by greggles** [comment]
There is an existing [antivirus](https://www.drupal.org/project/antivirus) module for Drupal which currently only supports ClamAV. I think the simplest way to get phpMussel integrated with Drupal's file handling would just be to make that module use either ClamAV or phpMussel. Also, I'm pretty sure the maintainers would be interested in supporting other tools beyond ClamAV, especially if they work on hosts where the clam command line is not available. I suggest opening an issue on Drupal.org to discuss it (go to https://www.drupal.org/project/issues/antivirus and when logged in click "create a new issue")  and then linking to the issue from this discussion will be the best way to get the right attention.

**Comment by DanielRuf** [comment]
@greggles I am not sure about this Drupal module as they are looking for a new maintainer and the last update was 1 year ago.

The ClamAV module is still maintained but has very few active committers and few updates.

I think a separate module for phpMussel with an own maintainer makes more sense.

**Comment by 737simpilot** [comment]
I believe I saw a user on phpbb that offered his extension writing skills for a fee. I might shoot him a message and see what he says. Mods are now discontinued as is 2.0. Its all 3.1.x now.

**Comment by Maikuolan** [comment]
@greggles Thanks for the tip. :-)

This could be a good idea, but, due to it being more than a year since their last update (commit, reply to an issue, etc), I suspect they likely wouldn't reply if I were to post there at the moment, but, the idea could be viable in the future, if a new maintainer took over and activity were to pick up.

I agree with @DanielRuf, though, in that phpMussel would be better as a separate module; This is because the majority (although not all) of the signatures that phpMussel uses come from ClamAV, anyhow (and the signatures that phpMussel uses is a smaller subset, due to the inherent limitations in memory and etc for any PHP-based anti-virus solution), and so, phpMussel would be redundant on a system that already entirely supports ClamAV. But, on any system that doesn't support any AV yet, nor is unable to install or support this sort of AV integration, phpMussel could be very valuable, due to being entirely PHP-based and not requiring external binaries and etc. So, probably best on a module of its own, or, where possible and where appropriate, bundled with other similar tools that can act as fallbacks for those systems where integration with external binaries isn't possible.

@737simpilot Thanks! It would be good if someone were interested. :-)

**Comment by 737simpilot** [comment]
Yeah, I think I know who does paid work for coding extensions. l'll see tonight for sure. It's 2 pm here right now.

**Comment by 737simpilot** [comment]
This is agent smith from spambot security,  btw

**Comment by Maikuolan** [comment]
Ah, cool. Good to see you here. :-)

**Comment by 737simpilot** [comment]
I created a topic here in the extension requests: https://www.phpbb.com/community/viewtopic.php?f=496&t=2351616&p=14304296#p14304296

I thought it was AJAX that prevented PHPMussel from working.

**Comment by DanielRuf** [comment]
phpMussel works but when AJAX is used the return value has to be processed by the CMS and display something for the user. Most CMS will not handle the custom answer from phpMussel (which is a 403 with a custom response).

This nothing directly to do with phpMussel or the CMS but how asynchronous JavaScript (AJAX) works.

Even if the CMS can not show the result, phpMussel still works in the background if properly hooked (best is the .htaccess solution, never change core files of a CMS or board software).

I think adding the hook to config.php is not enough. When an AJAX call with a malicious upload is not blocked then the called file is not using the config.php file. This is why I prefer the .htaccess solution. This makes sure that it is always hooked when the PHP interpreter is called eg when a php file is executed.

**Comment by Maikuolan** [comment]
Yep, this is correct. And so, ideally, it would be great to be able to find someone from phpBB that has a good understanding of the way that phpBB handles file uploads with its AJAX, to be interested in developing a phpMussel plugin for phpBB (because it would be easier for them to do than it would for someone lacking that understanding).

There's been some very helpful users that've given me some great tips and pointers as to how to go about doing it, and I'd definitely been considering doing it myself at some point in the future, failing all else, but it's something I hadn't yet gotten around to starting, and I do lack the aforementioned understanding, so, I'm still very keen to bring someone on-board that could do it for us, if we could find such an interested person.

Thanks for starting that topic there; Hopefully we'll get a few bites from it, maybe attract someone that might be interested. Appreciated. :-)

**Comment by DanielRuf** [comment]
@737simpilot any update? Would be great to have someone, who could create and maintain such a mod so we can focus on the other things.

**Comment by 737simpilot** [comment]
I have tried and tried. I have asked a guy who runs a website that apparently does coding for a fee, I have asked at phpbbmoddeers.net and their website has an SSL issue now. And I'm not getting anyone interested on the phpBB board. Looks like we are in this along which really suxx! Maikuolan, you may want to learn how to code extensions.

**Comment by DanielRuf** [comment]
Hm, so even the one who provides his coding srvices for a fee did not respond? We need someone who can develop and maintain this.

You can normally skip SSL warnings.

I've asked for some help in some board and groups today. Let's see if anyone is interested. Would also pay for this if we have someone, who would maintain this.

**Comment by 737simpilot** [comment]
Yep, nobody.

**Comment by Maikuolan** [comment]
Closing this issue now due to being stale (and I don't think any of us seem to really have enough time or motivation to follow up any further with it, myself included). Maybe we can create some new issues for specific, individual integrations for various CMS and such, if and when somebody feels that they've got the time and motivation for it.

**Comment by 737simpilot** [comment]
Just a quick note. The people at phpBB are general idiots in the realm of security and it's like pulling teeth to get someone to code an extension. I've been successful a few times, but mostly not. The one sheer fire way for me to do this is to post in the wanted section on the phpBB forum. Right now I'm strapped for cash and I just made a paid request for an extension I use to fix an issue since the author isn't making any progress. Yet when I requested professional paid help, the damn Dev chimed in and wanted money, yet his extension is free! Go figure. I could try my hand over at Fiverr as well.    But as I said, I'm strapped for cash. Maybe next month. I'd really like to have phpMussel to work with my phpBB board, but it errors out on upload after all these years testing it. Another factor is that I use an extension to upload attachments via Amazon AWS S3 so I hope that isn't a factor. It may or may not.
